### PR TITLE
Fix gestures recognizers end up in broken state after additional unaccounted for touch

### PR DESCRIPTION
**Pull request by cathyxz**
Fixes https://github.com/ampproject/amphtml/issues/17796. 
~1. Make sure that if a gesture returns false from `onTouchMove`, that we call `acceptCancel` to ensure that `this.eventing_` is set to false. This prevents the permanently broken state of `gestures`, where a recognizer is `eventing_` but not tracking anything and thus prevents subsequent gestures from being recognized. 
2. [Open to debate] Ignore all additional touches if a gesture is already eventing. I think this is correct behavior, but we can experiment with the UX.~

Update: 
We're going to go with the implementation here where if an additional touch is added while the recognizer is still eventing, it is ignored, and UX-wise the gesture will continue based on the first n touches that triggered the gesture. 

To recap, there were three possible solutions to this problem: 
1. Call `acceptCancel` on all return `false`. This would solve the problem where additional touches causes the eventing recognizer to stop listening to future gestures (and thus never stops eventing), but no future recognizers can trigger because there is already an eventing gesture. 
2. Early return true on additional touch (keep listening to subsequent events but ignore additional touches). This causes the gesture to pause while additional touches are present but to resume when the additional touches are gone. Results in slightly sub-optimal UX because users can unintentionally freeze the gesture due to a unnoticed extra touch. 
3. [Chosen approach] always emit on touch move even if there is an additional touch, but not to start the gesture if there are extra touches. UX-wise, it will ignore additional touch if a recognizer is already eventing, so if you put down or release a third finger while pinching or a second finger while swiping, the zoom / swipe continues as is based on your first 2 or 1 touch(es). 

Option 3 has the downside of this situation:  if you first move one finger to pan, then stop, then put second down, I can't pinch zoom. Similarly if you try to transition from a pinch zoom to a pan without lifting the finger, it doesn't register as a swipe. 

Demo link [here](https://amp-demo-project-1323a.firebaseapp.com/). 
Developed on Pixel 2 / Chrome, sanity checked on Safari.

### CODE REVIEW

File: src/gesture-recognizers.js [hunk 1]
```diff
@@ -280,6 +280,10 @@ class SwipeRecognizer extends GestureRecognizer {
   /** @override */
   onTouchStart(e) {
     const {touches} = e;
+    // If already eventing, ignore additional touches
+    if (this.eventing_ && touches && touches.length > 1) {
+      return true;
+    }
     if (touches && touches.length == 1) {
```
Comment (cathyxz) [inline src/gesture-recognizers.js]: There are two options here: 
1. Early return for additional touch case and continue listening to gesture. Slightly suboptimal UX because additional touch will result in gesture stopping until you lift the finger. 

```
    // If already eventing, ignore additional touches
    if (this.eventing_ && touches && touches.length > 1) {
      return true;
    }
```

2. Actually ignore the touch. 

```
if (touches && touches.length >= 1) 
```
This is better UX for a single gesture, but may result in confusion between swipe and pinch.

**Comment by cathyxz** [comment]
Actually hold off for now, I just realized only one of the two fixes are needed, not both.

File: src/gesture-recognizers.js [hunk 2]
```diff
@@ -762,18 +760,16 @@ export class PinchRecognizer extends GestureRecognizer {
...
       return true;
     }
-    // If already in the middle of a pinch event, ignore additional touches.
-    if (this.eventing_ && touches && touches.length > 2) {
-      return true;
-    }
     else if (touches && touches.length >= 2) {
```
Comment (sparhami) [inline src/gesture-recognizers.js]: Better to just turn this into an if-statement since the `if` is an early return.

File: test/functional/test-gesture-recognizers.js [hunk 3]
```diff
@@ -888,7 +888,7 @@ describe('PinchRecognizer', () => {
       {clientX: 120, clientY: 130}]});
     expect(res).to.equal(true);
 
-    gesturesMock.expects('signalEmit_').once();
+    gesturesMock.expects('signalEmit_').twice();
```
Comment (sparhami) [inline test/functional/test-gesture-recognizers.js]: Should add an `expect` to make sure that it used the correct touch (i.e. it used touches[0] and not the newly added touch).

File: src/gesture-recognizers.js [hunk 4]
```diff
@@ -312,15 +320,18 @@ class SwipeRecognizer extends GestureRecognizer {
           if (dx >= 8 && dx > dy) {
             this.signalReady(-10);
           } else if (dy >= 8) {
+            this.acceptCancel();
```
[the remark this answers is not in the document]
Comment (cathyxz) [reply src/gesture-recognizers.js]: In the most recent refactor, we won't travel down this code path if the recognizer is already eventing and will hence not need these `acceptCancel` calls.

File: src/gesture-recognizers.js (hunk 1 above)
Comment (cathyxz) [reply src/gesture-recognizers.js]: Confusion between swipe and pinch context here: https://github.com/ampproject/amphtml/pull/12219/files#r153873499. But I think this was already mitigated by adding a threshold to not trigger pinch if both fingers move in the same direction in `PinchRecognizer`'s `onTouchMove` function.

File: src/gesture-recognizers.js [hunk 5]
```diff
@@ -774,7 +793,11 @@ export class PinchRecognizer extends GestureRecognizer {
 
   /** @override */
   onTouchEnd(e) {
-    this.end_(e);
+    // Number of current touches on the page
+    const {touches} = e;
+    if (touches.length <= 1) {
```
[the remark this answers is not in the document]
Comment (sparhami) [reply src/gesture-recognizers.js]: The number 1 feels a bit weird. Feels like this might make more sense as `if (touches.length < 2)`, since 2 is the number of touches this gesture is interested in, so it seems to be more natural.

File: src/gesture-recognizers.js [hunk 6]
```diff
@@ -742,14 +756,17 @@ export class PinchRecognizer extends GestureRecognizer {
     // for second touch.
     if (touches && touches.length == 1) {
       return true;
-    } else if (touches && touches.length == 2) {
+    }
+    if (touches && touches.length >= 2) {
```
Comment (sparhami) [inline src/gesture-recognizers.js]: No need to change, but this feels a bit weird. Seems like if we don't have one touch, then either we have zero or two+. I'm not sure if you can have zero touches in a touchmove. I think this may flow better if it were:

```javascript
if (!touches || touches.length == 0) {
  return false;
}

if (touches.length == 1) {
  return true;
}

// Have 2+ touches
```

File: src/gesture.js [hunk 7]
```diff
@@ -297,7 +297,8 @@ export class Gestures {
         continue;
       }
       this.recognizers_[i].onTouchEnd(event);
-      if (!this.pending_[i] || this.pending_[i] < now) {
+      if ((!this.pending_[i] || this.pending_[i] < now)
+        && this.eventing_ != this.recognizers_[i]) {
```
Comment (sparhami) [inline src/gesture.js]: Perhaps pull this out into a variable to make it more clear. I think having a lot of complicated logic in an if-statement can get hard to follow. Something like:

```javascript
const isEventing = this.eventing_ == this.recognizers_[i];
if ((!this.pending_[i] || this.pending_[i] < now) && !isEventing) {
  this.stopTracking_(i);
}
```

or better one step further:

```javascript
const isEventing = this.eventing_ == this.recognizers_[i];
const isExpiredAfterEvent = !this.pending_[i] || this.pending_[i] < now;
if ((isExpiredAfterEvent && !isEventing) {
  this.stopTracking_(i);
}
```

I think this brings up a question for me: is it a problem that `this.pending_[i]` is < `now`? It seems like at this point, if another touchend were to happen (lets say you need 2 touches, you have 4, then you went to 3, then immediately to 2), then it seems like the if-statement above would call `stopTracking()`, even though you still have two touches.

This seems to indicate that a similar isEventing check might be needed above.

File: src/gesture-recognizers.js [hunk 8]
```diff
@@ -293,13 +297,15 @@ class SwipeRecognizer extends GestureRecognizer {
   /** @override */
   onTouchMove(e) {
     const {touches} = e;
-    if (touches && touches.length == 1) {
```
[the remark this answers is not in the document]
Comment (cathyxz) [reply src/gesture-recognizers.js]: Done.

File: src/gesture-recognizers.js [hunk 9]
```diff
@@ -723,6 +733,10 @@ export class PinchRecognizer extends GestureRecognizer {
     // for second touch.
     if (touches && touches.length == 1) {
       return true;
+    }
+    // If already in the middle of a pinch event, ignore additional touches.
+    if (this.eventing_ && touches && touches.length > 2) {
+      return true;
     } else if (touches && touches.length == 2) {
```
[the remark this answers is not in the document]
Comment (cathyxz) [reply src/gesture-recognizers.js]: Done

File: src/gesture-recognizers.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (cathyxz) [reply src/gesture-recognizers.js]: Done

File: src/gesture-recognizers.js [hunk 10]
```diff
@@ -332,7 +338,11 @@ class SwipeRecognizer extends GestureRecognizer {
 
   /** @override */
   onTouchEnd(e) {
-    this.end_(e);
+    const {touches} = e;
+    // Number of current touches on the page
+    if (touches.length == 0) {
```
[the remark this answers is not in the document]
Comment (cathyxz) [reply src/gesture-recognizers.js]: Done

File: src/gesture-recognizers.js (hunk 6 above)
Comment (cathyxz) [reply src/gesture-recognizers.js]: Refactored this function for clarity. PTAL.

File: src/gesture-recognizers.js [hunk 11]
```diff
@@ -738,43 +756,78 @@ export class PinchRecognizer extends GestureRecognizer {
...
+    }
+    // Pinch gesture detected but threshold not reached, continue listening
+    return true;
+  }
+
+  /**
+   * @return {boolean}
+   * @private
+   */
+  pinchReady_() {
```
Comment (sparhami) [inline src/gesture-recognizers.js]: isPinchReady_?

File: src/gesture.js (hunk 7 above)
Comment (cathyxz) [reply src/gesture.js]: Per discussion offline, pulled out the booleans into separate named constants. The previous block should not have an eventing check because of the `continue`. If it's eventing, then we definitely want to call the recognizer's `onTouchEnd`, not skip it. I think that would be a larger refactor that would be the scope for a different PR.

File: src/gesture-recognizers.js (hunk 11 above)
Comment (cathyxz) [reply src/gesture-recognizers.js]: Done and also fixed below. Technically, it's more like "pinchIsReady", "pinchIsRejected". Or maybe since these are private functions of the `PinchRecognizer`, we could directly go with `isReady` and `isRejected`.

Comment (cathyxz) [reply src/gesture-recognizers.js]: Talked offline, going to leave as is (`isPinchReady`, `isPinchRejected`).
